### PR TITLE
[inductor] Set CUDA_VISIBLE_DEVICES for multi-device subprocess autotuning

### PR DESCRIPTION
**Pull request by masnesral**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109500

Summary: The curent parallel autotune implementation sets the CUDA_VISIBLE_DEVICES env var too late -- after the benchmarking subprocess has started -- and the torch libraries don't recognize the change. Since the multiprocessing library doesn't support providing an environment for the subprocess, temporarily set CUDA_VISIBLE_DEVICES in the parent process so that the change is inherited by the subprocess.

Test Plan:
* New unit test to verify the env var is set in the sub-process and fail the benchmark if it's not.
* Ran multiprocess autotuning and looked at the output from `nvidia-smi pmon` to make sure that all GPUs were assigned processes.

Snippet:
```
    1    3442314     C     2     1     -     -   python
    2    3442318     C     2     1     -     -   python
    3    3442320     C     8     2     -     -   python
    4    3442323     C     9     4     -     -   python
    5    3442325     C    10     4     -     -   python
    6    3442327     C    10     4     -     -   python
    7    3442329     C     2     0     -     -   python
    0    3434906     C     0     0     -     -   python
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/autotune_process.py [hunk 1]
```diff
@@ -42,6 +43,27 @@ class Pong:
...
 
 
+@contextlib.contextmanager
+def set_cuda_visible_device(device: Optional[int]):
+    """
+    Context manager to set the CUDA_VISIBLE_DEVICES environment variable to the
+    specified single device. If device is None, don't manipulate the environment.
+    """
+    if device is None:
+        yield
```
Comment (aakhundov) [inline torch/_inductor/autotune_process.py:53]: To double check my understanding: if the user sets `CUDA_VISIBLE_DEVICES=7` on the parent python process running `torch.compile` and multi-device autotuning is *disabled*, we follow this code path, but the env var value should propagate to the child process. Is this correct?

Comment (aakhundov) [reply torch/_inductor/autotune_process.py:53]: Thanks for the detailed explanation! This looks correct.

File: test/inductor/test_max_autotune.py [hunk 2]
```diff
@@ -487,28 +493,51 @@ def __str__(self) -> str:
...
+            choice.bmreq.valid_devices = (None,)
 
             timings = tuning_pool.benchmark([choice])
             self.assertTrue(choice in timings)
-            self.assertEqual(timings[choice], value)
+            self.assertEqual(timings[choice], bmreq.value)
+
+            tuning_pool.terminate()
+
+    def test_tuning_pool_multiple_devices(self):
```
Comment (eellison) [inline test/inductor/test_max_autotune.py:523]: You might need `requires_multigpu` decorator here

**Review by eellison — APPROVED** [review]
looks good !

File: torch/testing/_internal/inductor_utils.py [hunk 3]
```diff
@@ -9,7 +9,9 @@
 from torch._dynamo.backends.registry import register_backend
 from torch._inductor.compile_fx import compile_fx, count_bytes_inner
 
+import functools
```
Comment (masnesral) [inline torch/testing/_internal/inductor_utils.py:12]: @eellison, mind taking a quick look at the new changes to this file?

**Review by shunting314 — APPROVED** [review]

**Pull request closed without merging**